kubetest2/kubetest2-gce/deployer/common.go

```diff
@@ -0,0 +1,162 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deployer
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"k8s.io/klog"
+	"sigs.k8s.io/boskos/client"
+	boskosCommon "sigs.k8s.io/boskos/common"
+)
+
+func (d *deployer) init() error {
+	if d.commonOptions.ShouldBuild() {
+		if err := d.verifyBuildFlags(); err != nil {
+			return fmt.Errorf("init failed to check build flags: %s", err)
+		}
+	}
+
+	if d.commonOptions.ShouldUp() {
+		if err := d.verifyUpFlags(); err != nil {
+			return fmt.Errorf("init failed to verify flags for up: %s", err)
+		}
+
+		if d.GCPProject == "" {
+			klog.Info("No GCP project provided, acquiring from Boskos")
+
+			if err := d.getProjectFromBoskos(); err != nil {
+				return fmt.Errorf("init failed to get project from boskos: %s", err)
+			}
+			klog.Infof("Got project %s from boskos", d.GCPProject)
+		}
+
+	}
+
+	if d.commonOptions.ShouldDown() {
+		if err := d.verifyDownFlags(); err != nil {
+			return fmt.Errorf("init failed to verify flags for down: %s", err)
+		}
+	}
+
+	return nil
+}
+
+// getProjectFromBoskos creates a boskos client, acquires a gcp project
+// and starts a heartbeat goroutine to keep the project reserved
+func (d *deployer) getProjectFromBoskos() error {
+	boskos, err := client.NewClient(
+		os.Getenv("JOB_NAME")+"-kubetest2",
+		d.BoskosLocation,
+		"",
+		"",
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create boskos client: %s", err)
+	}
+
+	resourceType := "gce-project"
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute*5)
+	defer cancel()
+
+	boskosProject, err := boskos.AcquireWait(ctx, resourceType, "free", "busy")
+	if err != nil {
+		return fmt.Errorf("failed to get a %s from boskos: %s", resourceType, err)
+	}
+	if boskosProject == nil {
+		return fmt.Errorf("boksos had no %s available", resourceType)
+	}
+
+	go func(c *client.Client, resource *boskosCommon.Resource) {
+		for range time.Tick(time.Minute * 5) {
+			if err := c.UpdateOne(resource.Name, "busy", nil); err != nil {
+				klog.Warningf("[Boskos] Update of %s failed with %v", resource.Name, err)
+			}
+		}
+	}(boskos, boskosProject)
+
+	d.boskos = boskos
+	d.boskosProject = boskosProject
+	d.GCPProject = boskosProject.Name
+
+	return nil
+}
+func (d *deployer) verifyUpFlags() error {
+	if err := d.setRepoPathIfNotSet(); err != nil {
+		return err
+	}
+
+	d.kubectl = filepath.Join(d.RepoRoot, "cluster", "kubectl.sh")
+
+	// verifyUpFlags does not check for a gcp project because it is
+	// assumed that one will be acquired from boskos if it is not set
+	if d.BoskosLocation == "" {
+		d.BoskosLocation = "http://boskos.test-pods.svc.cluster.local."
```

Comment:
consider making this the flag default value in general.

Reply:
Done.